peps/pep-0776.rst

```diff
@@ -0,0 +1,794 @@
+ignored (`see step 7 here
+<https://webassembly.github.io/spec/js-api/index.html#call-an-exported-function)>`__.
+This works, but has the disadvantage of being slow and breaking stack switching
+-- it is not possible to stack switch through JavaScript frames.
+
+Using the wasm-gc `ref.test
+<https://webassembly.github.io/gc/core/exec/instructions.html#xref-syntax-instructions-syntax-instr-ref-mathsf-ref-test-mathit-rt>`__
+instruction, we can query the type of the function pointer and manually fix up
+the argument list.
+
+wasm-gc is a relatively new feature for WebAssembly runtimes, so we attempt to
+use a wasm-gc based function pointer cast trampoline if possible and fall back
+to a JS trampoline if not. Every JavaScript runtime that supports stack
+switching also supports wasm-gc, so this ensures that stack switching works on
+every platform runtime that supports it. The one wrinkle is that iOS 18 ships a
+broken implementation of wasm-gc so we have to special case it.
+
+`See here for the full implementation details.
+<https://github.com/python/cpython/blob/98fa4a49fecbac3c990a25ce5d300592dad31be0/Python/emscripten_trampoline.c>`__
+
+The function pointer cast handling is fully implemented in cpython. Pyodide uses
+exactly the same code as upstream.
+
+
+------------
+CI Resources
+------------
+
+Pyodide can be built and tested on any Linux with a reasonably recent version of
+Node. Anaconda has offered to provide physical hardware to run Android
+buildbots, maintained by Russell Keith-Magee.
+
+CPython does not currently test Tier 3 platforms on GitHub Actions, but if this
+ever changes, their Linux runners are able to build and test Emscripten Python.
+
+---------
+Packaging
+---------
+
+Existing Package Support
+========================
+
+Pyodide currently maintains ports of 255 different packages at the time of this
+writing, including major scientific Python packages like numpy, scipy, pandas,
+polars, scikit-learn, opencv, pyarrow, and Pillow as well as general purpose
+packages like aiohttp, requests, pydantic, cryptography, and orjson.
+
+About 60 packages are also testing against Pyodide in their CI, including numpy,
+pandas, awkward-cpp, scikit-image, statsmodels, pyarrow, hypothesis, and PyO3.
+
+Emscripten Wheel Format
+=======================
+
+Emscripten wheels will use either the format ``emscripten_<version>_wasm32`` or
+``pyodide_<abi>_wasm32``. For example:
+
+* ``emscripten_3_1_58_wasm32``
+* ``pyodide_2025_0_wasm32``
+
+The first triple is ambiguous, since even with Emscripten 3.1.58 it is possible
+to link dynamic libraries that require a large number of distinct ABIs,
+depending on linker and compiler options. It is our intent that the
+``pyodide_2025_0`` specifies the particular ABI. Thus, the relationship between
+``pyodide_<abi>`` and ``emscripten_<version>`` is intended to be the same as the
+relationship between ``manylinux<version>`` and ``linux``.
+
+The specification of the ``pyodide_<abi>`` ABI includes:
```

Comment:
It would be good if there was a link to some page where these ABI versions are going to be tracked maybe?

Reply:
I've been meaning to add to Pyodide a `policies` page, which would be a good place to refer to for this. I think the most contentious part of this PEP is likely to be who/what organization chooses the ABI and what it is called...

Reply:
I've commented on this on the discussion thread; but tl;dr, I've got concerns about using `pyodide_<abi>` as the naming here, even if in practice we end up using Pyodide's specification.